src/Xamarin.Android.Build.Tasks/Utilities/NativeLinker.cs

```diff
@@ -0,0 +1,430 @@
+	{
+		// The exception is just a precaution, since the items passed to us should have already been checked
+		string itemAbi = item.GetMetadata (KnownMetadata.Abi) ?? throw new InvalidOperationException ($"Internal error: 'Abi' metadata not found in item '{item}'");
+		if (String.Compare (abi, itemAbi, StringComparison.OrdinalIgnoreCase) == 0) {
+			return;
+		}
+
+		throw new InvalidOperationException ($"Internal error: '{item}' ABI ('{itemAbi}') doesn't have the expected value '{abi}'");
+	}
+
+	void EnsureCorrectAbi (List<ITaskItem>? items)
+	{
+		if (items == null) {
+			return;
+		}
+
+		foreach (ITaskItem item in items) {
+			EnsureCorrectAbi (item);
+		}
+	}
+
+	bool ExtractDebugSymbols (ITaskItem outputSharedLibrary)
+	{
+		var stdoutLines = new List<string> ();
+		var stderrLines = new List<string> ();
+
+		string sourceLib = outputSharedLibrary.ItemSpec;
+		string sourceLibQuoted = MonoAndroidHelper.QuoteFileNameArgument (sourceLib);
+		string destLib = Path.Combine (Path.GetDirectoryName (sourceLib), $"{Path.GetFileNameWithoutExtension (sourceLib)}.dbg.so");
+		string destLibQuoted = MonoAndroidHelper.QuoteFileNameArgument (destLib);
+
+		var args = new List<string> {
+			"--only-keep-debug",
+			sourceLibQuoted,
+			destLibQuoted,
+		};
+
+		if (!RunCommand ("Extract Debug Info", objcopy, args, stdoutLines, stderrLines)) {
+			LogFailure ();
+			return false;
+		}
+
+		stdoutLines.Clear ();
+		stderrLines.Clear ();
+
+		args.Clear ();
+		args.Add ("--strip-debug");
+		args.Add ("--strip-unneeded");
+		args.Add (sourceLibQuoted);
+
+		if (!RunCommand ("Strip Debug Info", objcopy, args, stdoutLines, stderrLines)) {
+			LogFailure ();
+			return false;
+		}
+
+		stdoutLines.Clear ();
+		stderrLines.Clear ();
+		args.Clear ();
+		args.Add ($"--add-gnu-debuglink={destLibQuoted}");
+		args.Add (sourceLibQuoted);
+
+		if (!RunCommand ("Add Debug Info Link", objcopy, args, stdoutLines, stderrLines)) {
+			LogFailure ();
+			return false;
+		}
+
+		return true;
+
+		void LogFailure ()
+		{
+			var sb = MonoAndroidHelper.MergeStdoutAndStderrMessages (stdoutLines, stderrLines);
+			// TODO: consider making it a warning
+			// TODO: make it a coded message
+			log.LogError ("Failed to extract debug info", Path.GetFileName (sourceLib), sb.ToString ());
+		}
+	}
+
+	bool RunLinker (List<string> args, ITaskItem outputSharedLibrary)
+	{
+		var stdoutLines = new List<string> ();
+		var stderrLines = new List<string> ();
+
+		if (!RunCommand ("Native Linker", ld, args, stdoutLines, stderrLines)) {
+			var sb = MonoAndroidHelper.MergeStdoutAndStderrMessages (stdoutLines, stderrLines);
+			log.LogCodedError ("XA3007", Properties.Resources.XA3007, Path.GetFileName (outputSharedLibrary.ItemSpec), sb.ToString ());
+			return false;
+		}
+
+		return true;
+	}
+
+	bool RunCommand (string label, string binaryPath, List<string> args, List<string> stdoutLines, List<string> stderrLines)
+	{
+		using var stdout_completed = new ManualResetEvent (false);
```

Comment:
MonoAndroidHelper has a `RunProcess` static, can that be used for this?

Reply:
It could, sure, but not without some changes: warning in `MonoAndroidHelper` would have to be parameterized and cancellation token support added. It's a good idea, I'll make the necessary changes.